src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -782,7 +782,7 @@ private HeaderStoreItemInfo CreateAndAddHeaderToStore(HeaderDescriptor descripto
             // If we don't have the header in the store yet, add it now.
             HeaderStoreItemInfo result = new HeaderStoreItemInfo();
 
-            AddHeaderToStore(descriptor, result);
+            AddHeaderToStore((descriptor.HeaderType & _treatAsCustomHeaderTypes) != 0 ? descriptor.AsCustomHeader() : descriptor, result);
```

Comment:
I'm not sure this is the right place for this fix.

In most cases, the descriptor passed in here will already have been checked against _treatAsCustomHeaderTypes and converted to a "custom header" descriptor if necessary. So doing this here is redundant.

The only case I can see where this isn't being done is the case where we are parsing response headers in HttpConnection and calling HeaderDescriptor.TryGet directly. (And the equivalent in Http2Stream for HTTP2.)

We should probably just modify the logic in HttpConnection to ensure it's doing the same "custom header" handling. That is, check if it's a request header and if so, convert it to a "custom header" descriptor.

It would be good to add an assert here (in CreateAndAddHeaderToStore) as well.

I kind of wish we had a type-level distinction between "raw HeaderDescriptor" and "HeaderDescriptor appropriate for this header collection type (i.e. converted to custom if necessary, content headers only allowed on content header collections)" but that's a whole other can of worms.

Reply:
Ok, I added the assertion and moved the logic into `HttpConnection`. It seemed a bit weird to reference `HeaderType.Request` directly, but `_treatAsCustomHeaderTypes` was not exposed anywhere, and it was always going to be set to `HeaderType.Request`. :)

Reply:
It's a little weird, but we're comparing against HeaderType.Content just above already. This code basically needs to understand the difference between header types.